File: library/src/main/java/com/chuckerteam/chucker/api/Chucker.java
/*
 * Copyright (C) 2017 Jeff Gilfelt.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.chuckerteam.chucker.api;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.IntDef;

import com.chuckerteam.chucker.api.internal.support.ChuckerCrashHandler;
import com.chuckerteam.chucker.api.internal.support.NotificationHelper;
import com.chuckerteam.chucker.api.internal.ui.MainActivity;

/**
 * Chucker utilities.
 */
public class Chucker {

    public static final int SCREEN_HTTP = 1;
    public static final int SCREEN_ERROR = 2;

    /**
     * Get an Intent to launch the Chucker UI directly.
     *
     * @param context A Context.
     * @param screen The screen to display: SCREEN_HTTP or SCREEN_ERROR.
     * @return An Intent for the main Chucker Activity that can be started with {@link Context#startActivity(Intent)}.
     */
    public static Intent getLaunchIntent(Context context, @Screen int screen) {
        return new Intent(context, MainActivity.class)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                .putExtra(MainActivity.EXTRA_SCREEN, screen);
    }

    /**
     * Configure the default crash handler of the JVM to report all uncaught Throwable to Chucker.
     * You may only use it for debugging purpose.
     *
     * @param collector the ChuckerCollector
     */
    public static void registerDefaultCrashHanlder(final ChuckerCollector collector) {
        Thread.setDefaultUncaughtExceptionHandler(new ChuckerCrashHandler(collector));
    }

    public static void dismissTransactionsNotification(Context context) {
        new NotificationHelper(context).dismissTransactionsNotification();
    }

    public static void dismissErrorsNotification(Context context) {
        new NotificationHelper(context).dismissErrorsNotification();
    }

    /**
     * Check if this instance is the operation one or no-op.
     *
     * @return {@code true} if this is the operation instance.
     */
    public static boolean isOp() {
        return true;
    }

    @IntDef(value = {SCREEN_HTTP, SCREEN_ERROR})
    public @interface Screen {}
}

File: library/src/main/kotlin/com/chuckerteam/chucker/api/Chucker.kt
package com.chuckerteam.chucker.api

import android.content.Context
import android.content.Intent
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.graphics.drawable.Icon
import android.net.Uri
import android.os.Build
import android.util.Log
import android.widget.Toast
import androidx.core.content.getSystemService
import com.chuckerteam.chucker.R
import com.chuckerteam.chucker.internal.data.repository.RepositoryProvider
import com.chuckerteam.chucker.internal.support.Logger
import com.chuckerteam.chucker.internal.support.NotificationHelper
import com.chuckerteam.chucker.internal.support.TransactionListDetailsSharable
import com.chuckerteam.chucker.internal.support.writeToFile
import com.chuckerteam.chucker.internal.ui.MainActivity
import kotlinx.coroutines.runBlocking

/**
 * Chucker methods and utilities to interact with the library.
 */
public object Chucker {

    private const val SHORTCUT_ID = "chuckerShortcutId"

    /**
     * Check if this instance is the operation one or no-op.
     * @return `true` if this is the operation instance.
     */
    @Suppress("MayBeConst ") // https://github.com/ChuckerTeam/chucker/pull/169#discussion_r362341353
    public val isOp: Boolean = true

    /**
     * Get an Intent to launch the Chucker UI directly.
     * @param context An Android [Context].
     * @return An Intent for the main Chucker Activity that can be started with [Context.startActivity].
     */
    @JvmStatic
    public fun getLaunchIntent(context: Context): Intent {
        return Intent(context, MainActivity::class.java)
            .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
    }

    /**
     * Create a shortcut to launch Chucker UI.
     * @param context An Android [Context].
     */
    internal fun createShortcut(context: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N_MR1) {
            return
        }

        val shortcutManager = context.getSystemService<ShortcutManager>() ?: return
        if (shortcutManager.dynamicShortcuts.any { it.id == SHORTCUT_ID }) {
            return
        }

        val shortcut = ShortcutInfo.Builder(context, SHORTCUT_ID)
            .setShortLabel(context.getString(R.string.chucker_shortcut_label))
            .setLongLabel(context.getString(R.string.chucker_shortcut_label))
            .setIcon(Icon.createWithResource(context, R.mipmap.chucker_ic_launcher))
            .setIntent(getLaunchIntent(context).setAction(Intent.ACTION_VIEW))
            .build()
        try {
            shortcutManager.addDynamicShortcuts(listOf(shortcut))
        } catch (e: IllegalArgumentException) {
            Logger.warn("ShortcutManager addDynamicShortcuts failed ", e)
        } catch (e: IllegalStateException) {
            Logger.warn("ShortcutManager addDynamicShortcuts failed ", e)
        }
    }

    /**
     * Dismisses all previous Chucker notifications.
     */
    @JvmStatic
    public fun dismissNotifications(context: Context) {
        NotificationHelper(context).dismissNotifications()
    }

    internal var logger: Logger = object : Logger {
        val TAG = "Chucker"

        override fun info(message: String, throwable: Throwable?) {
            Log.i(TAG, message, throwable)
        }

        override fun warn(message: String, throwable: Throwable?) {
            Log.w(TAG, message, throwable)
        }

        override fun error(message: String, throwable: Throwable?) {
            Log.e(TAG, message, throwable)
        }
    }
}
